src/core/border_router/routing_manager.cpp

```diff
@@ -3689,6 +3709,64 @@ void RoutingManager::PdPrefixManager::ProcessPlatformGeneratedRa(const uint8_t *
     }
 }
 
+void RoutingManager::PdPrefixManager::ProcessDhcpPdPrefix(const PrefixTableEntry &aPrefixTableEntry)
+{
+    bool                         currentPrefixUpdated = false;
+    DiscoveredPrefixTable::Entry entry;
+
+    VerifyOrExit(mEnabled, LogWarn("Ignore delegated prefix since PD is disabled."));
+
+    entry.SetFrom(aPrefixTableEntry);
+
+    if (!IsValidPdPrefix(entry.GetPrefix()))
+    {
+        LogWarn("PdPrefixManager: Ignore invalid received prefix %s", entry.GetPrefix().ToString().AsCString());
+        goto exit;
+    }
+
+    entry.mPrefix.SetLength(kOmrPrefixLength);
```

Comment:
This method shares a lot code with `PdPrefixManager::Process()`. I wonder if we refactor them to avoid repeated code?

Reply:
The `Process()` gets a RA message which may contain more than one prefix (so it has a loop though all PIO options to process each one). In the new method we get one prefix. 

Some ideas:
- Can update `Process()` to get both inputs as pointer (an RA message or a `PrefixTableEntry`) with one being provided and other one `null`. ```Process(const RouterAdvert::RxMessage *aMessage, const PrefixTableEntry *aPrefixTableEntry)```
-  Probably need to refactor the code that processes an `entry` into its own small method:
```c++
        if (!IsValidPdPrefix(entry.GetPrefix()))
        {
            LogWarn("PdPrefixManager: Ignore invalid PIO entry %s", entry.GetPrefix().ToString().AsCString());
            continue;
        }

        entry.mPrefix.Tidy();
        entry.mPrefix.SetLength(kOmrPrefixLength);

        // The platform may send another RA message to announce that the current prefix we are using is no longer
        // preferred or no longer valid.
        if (entry.GetPrefix() == GetPrefix())
        {
            currentPrefixUpdated = true;
            mPrefix              = entry;
        }

        if (entry.IsDeprecated())
        {
            continue;
        }

        // Some platforms may delegate us more than one prefixes. We will pick the smallest one. This is a simple rule
        // to pick the GUA prefix from the RA messages since GUA prefixes (2000::/3) are always smaller than ULA
        // prefixes (fc00::/7).
        if (favoredEntry.GetPrefix().GetLength() == 0 || entry.GetPrefix() < favoredEntry.GetPrefix())
        {
            favoredEntry = entry;
        }
```

Reply:
I have updated the code to merge both cases into `Process`